lib/Dialect/Polynomial/Conversions/PolynomialToModArith/PolynomialToModArith.cpp

```diff
@@ -0,0 +1,456 @@
+  using OpAdaptor = typename SourceOp::Adaptor;
+
+  ConvertCommon(SourceOp op, OpAdaptor adaptor,
+                const TypeConverter *typeConverter) {
+    // all current implemented Op use result as the source of polynomial type
+    resultType = op.getResult().getType();
+    polynomialType = cast<PolynomialType>(resultType);
+
+    ringAttr = polynomialType.getRing();
+
+    coefficientType = ringAttr.getCoefficientType();
+    coefficientTypeWidth = coefficientType.getIntOrFloatBitWidth();
+    coefficientIntegerType = dyn_cast<IntegerType>(coefficientType);
+    coefficientFloatType = dyn_cast<FloatType>(coefficientType);
+    // coefficient natural modulus is the power-of-two modulus
+    // e.g. 66 for i64 as one for sign bit and one for 2 ** 64
+    coefficientNaturalModulus =
+        APInt::getOneBitSet(coefficientTypeWidth + 2, coefficientTypeWidth);
+
+    coefficientModulus = ringAttr.getCoefficientModulus();
+    polynomialModulus = ringAttr.getPolynomialModulus();
+
+    coefficientModulusIsNotNatural = false;
+    if (coefficientModulus) {
+      coefficientModulusValue = coefficientModulus.getValue();
+      coefficientModulusWidth = coefficientModulusValue.getActiveBits();
+      // ringAttr verifier requires coefficientModulus <=
+      // coefficientNaturalModulus
+      coefficientModulusIsNotNatural =
+          coefficientModulusWidth <= coefficientTypeWidth;
+    } else if (coefficientIntegerType) {
+      // if modulus is not specified, lowering pass can choose
+      // to naturally wrap around the underlying type
+      coefficientModulusValue = coefficientNaturalModulus;
+      coefficientModulusWidth = coefficientModulusValue.getActiveBits();
+    }
+
+    // tensor related
+    resultTypeConverted = typeConverter->convertType(resultType);
+    resultTensorType = cast<RankedTensorType>(resultTypeConverted);
+    resultShape = resultTensorType.getShape()[0];
+    resultElementType = resultTensorType.getElementType();
+    resultElementWidth = resultElementType.getIntOrFloatBitWidth();
+    resultElementIntegerType = dyn_cast<IntegerType>(resultElementType);
+    resultElementFloatType = dyn_cast<FloatType>(resultElementType);
+
+    llvm::TypeSwitch<Operation &>(*op)
+        .Case<FromTensorOp>(
+            [&](auto op) { inputType = op.getOperand().getType(); })
+        .template Case<AddOp, SubOp, MulScalarOp>(
+            [&](auto op) { inputType = op.getOperands()[0].getType(); });
+    if (inputType) {
+      inputTypeConverted = typeConverter->convertType(inputType);
+      inputTensorType = cast<RankedTensorType>(inputTypeConverted);
+      inputShape = inputTensorType.getShape()[0];
+      inputElementType = inputTensorType.getElementType();
+      inputElementWidth = inputElementType.getIntOrFloatBitWidth();
+      inputElementIntegerType = dyn_cast<IntegerType>(inputElementType);
+      inputElementFloatType = dyn_cast<FloatType>(inputElementType);
+    }
+  }
+
+  // polynomial type related
```

Comment:
We don't use some of the following variables outside of the above function. Do we anticipate using them in the future? Otherwise I think it would best to keep them local.

Reply:
for input* (e.g. inputTensor, inputElementWidth) it is indeed FromTensor specific and now moved to ConvertFromTensor. Some other unused variable are also eliminated, and I think the current variables are essential and may be used in the future.